From 203a9270d6dafb0c739665f15ae55588ae52510d Mon Sep 17 00:00:00 2001
From: Jean-Christophe Trotin <jean-christophe.trotin@st.com>
Date: Wed, 7 Sep 2016 11:57:36 +0200
Subject: [PATCH 10/23] v4l2object: disable copy threshold if "memory:DMABuf"
 in caps

If DMABuf memory is used (0-copy), then the copy threshold feature
shall be disabled: allocating more buffers is not allowed in this
case.

Change-Id: I0f0430f1694e7ad34007608caac32f5c756b6370
Reviewed-on: https://gerrit.st.com/57582
Reviewed-by: Jean Christophe TROTIN <jean-christophe.trotin@st.com>
Tested-by: Jean Christophe TROTIN <jean-christophe.trotin@st.com>
Reviewed-by: Benjamin GAIGNARD <benjamin.gaignard@st.com>
Reviewed-by: Vincent ABRIOU <vincent.abriou@st.com>
---
 sys/v4l2/gstv4l2object.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index e5a59df..8bd8b07 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -43,6 +43,7 @@
 #include "gst/gst-i18n-plugin.h"
 
 #include <gst/video/video.h>
+#include <gst/allocators/allocators.h>
 
 GST_DEBUG_CATEGORY_EXTERN (v4l2_debug);
 #define GST_CAT_DEFAULT v4l2_debug
@@ -3983,8 +3984,14 @@ gst_v4l2_object_decide_allocation (GstV4l2Object * obj, GstQuery * query)
      * buffers and enable copy threshold */
     if (!update) {
       own_min += 3;
-      gst_v4l2_buffer_pool_copy_at_threshold (GST_V4L2_BUFFER_POOL (pool),
-          TRUE);
+      if (!gst_caps_features_contains (gst_caps_get_features (caps, 0),
+              GST_CAPS_FEATURE_MEMORY_DMABUF)) {
+        gst_v4l2_buffer_pool_copy_at_threshold (GST_V4L2_BUFFER_POOL (pool),
+            TRUE);
+      } else {
+        gst_v4l2_buffer_pool_copy_at_threshold (GST_V4L2_BUFFER_POOL (pool),
+            FALSE);
+      }
     } else {
       gst_v4l2_buffer_pool_copy_at_threshold (GST_V4L2_BUFFER_POOL (pool),
           FALSE);
-- 
2.7.4

